explore/marginal-gain/1.Marginal-Gain-TP53-Example.py

```diff
@@ -0,0 +1,327 @@
+    ], axis='columns')
+    display(cv_result_df[k].head(2))
+
+
+# In[170]:
+
+# Cross-validated performance heatmap
+for i,k in enumerate(['a','b']):
+    ax = plt.subplot(2,1,i+1)
+    cv_score_mat = pd.pivot_table(cv_result_df[k], values='mean_test_score', index='classify__l1_ratio', columns='classify__alpha')
+
+    ax = sns.heatmap(cv_score_mat, annot=True, fmt='.1%')
+    if i == 1: ax.set_xlabel('Regularization strength multiplier (alpha)')
+    else: ax.set_xlabel('')
+    ax.set_ylabel('Elastic net mix param (l1_ratio)');
+
+
+# ## Use Optimal Hyperparameters to Output ROC Curve
+
+# In[180]:
+
+def get_threshold_metrics(y_true, y_pred):
+    roc_columns = ['fpr', 'tpr', 'threshold']
+    roc_items = zip(roc_columns, roc_curve(y_true, y_pred))
+    roc_df = pd.DataFrame.from_items(roc_items)
+    auroc = roc_auc_score(y_true, y_pred)
+    return {'auroc': auroc, 'roc_df': roc_df}
+
+y_pred_train = {}
+y_pred_test = {}
+metrics_train = {}
+metrics_test = {}
+for k in ['a','b']:
+    y_pred_train[k] = cv_pipeline[k].decision_function(X_train[k])
+    y_pred_test[k] = cv_pipeline[k].decision_function(X_test[k])
+
+    metrics_train[k] = get_threshold_metrics(y_train, y_pred_train[k])
+    metrics_test[k] = get_threshold_metrics(y_test, y_pred_test[k])
+
+
+# In[186]:
+
+# Plot ROC
+for i,k in enumerate(['a','b']):
+    for label, metrics in ('Training', metrics_train[k]), ('Testing', metrics_test[k]):
+        roc_df = metrics['roc_df']
+        plt.plot(roc_df.fpr, roc_df.tpr,
+            label='{} (AUROC = {:.1%})'.format(label+'_'+str(k), metrics['auroc']))
+    plt.xlim([0.0, 1.0])
+    plt.ylim([0.0, 1.05])
+    plt.xlabel('False Positive Rate')
+    plt.ylabel('True Positive Rate')
+    plt.title('Predicting TP53 mutation from gene expression (ROC curves)')
+    plt.legend(loc='lower right');
+
+
+# ## What are the classifier coefficients?
+
+# In[188]:
+
+final_pipeline = {}
+final_classifier = {}
+for k in ['a','b']:
+    final_pipeline[k] = cv_pipeline[k].best_estimator_
+    final_classifier[k] = final_pipeline[k].named_steps['classify']
+
+
+# In[194]:
+
+coef_df = {}
+for k in ['a','b']:
+    select_indices = final_pipeline[k].named_steps['select'].transform(
+        np.arange(len(X[k].columns)).reshape(1, -1)
+    ).tolist()
+
+    coef_df[k] = pd.DataFrame.from_items([
+        ('feature', X[k].columns[select_indices]),
+        ('weight', final_classifier[k].coef_[0]),
+    ])
+
+    coef_df[k]['abs'] = coef_df[k]['weight'].abs()
+    coef_df[k] = coef_df[k].sort_values('abs', ascending=False)
+
+
+# In[196]:
+
+for k in ['a','b']:
+    print('{:.1%} zero coefficients; {:,} negative and {:,} positive coefficients'.format(
+            (coef_df[k].weight == 0).mean(),
+            (coef_df[k].weight < 0).sum(),
+            (coef_df[k].weight > 0).sum()
+    ))
+
+
+# In[263]:
+
+# What are the top weighted features for model a and model b?
+display(coef_df['b'].head(5))
```

Comment:
Can we add `n_positives` and `n_negatives` to these dataframes? Would help me diagnose the variability in performance between CV, training, & testing.